java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/HashTableDictionaryEncoder.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.ElementAddressableVector;
+
+/**
+ * Dictionary encoder based on hash table.
+ * @param <E> encoded vector type.
+ * @param <D> decoded vector type, which is also the dictionary type.
+ */
+public class HashTableDictionaryEncoder<E extends BaseIntVector, D extends ElementAddressableVector> {
+
+  /**
+   * The dictionary for encoding/decoding.
+   * It must be sorted.
+   */
+  private final D dictionary;
+
+  /**
+   * The hasher used to compute the hash code.
+   */
+  private final ArrowBufHasher hasher;
+
+  /**
+   * A flag indicating if null should be encoded.
+   */
+  private final boolean encodeNull;
+
+  /**
+   * The hash map for distinct dictionary entries.
+   * The key is the pointer to the dictionary element, whereas the value is the index in the dictionary.
+   */
+  private HashMap<ArrowBufPointer, Integer> hashMap = new HashMap<>();
+
+  /**
+   * The pointer used to probe each element to encode.
+   */
+  private ArrowBufPointer reusablePointer;
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary.
+   *
+   */
+  public HashTableDictionaryEncoder(D dictionary) {
+    this(dictionary, false);
+  }
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary.
+   * @param encodeNull a flag indicating if null should be encoded.
+   *     It determines the behaviors for processing null values in the input during encoding/decoding.
+   *    <li>
+   *       For encoding, when a null is encountered in the input,
+   *       1) If the flag is set to true, the encoder searches for the value in the dictionary,
+   *       and outputs the index in the dictionary.
+   *       2) If the flag is set to false, the encoder simply produces a null in the output.
+   *    </li>
+   *    <li>
+   *       For decoding, when a null is encountered in the input,
+   *       1) If the flag is set to true, the decoder should never expect a null in the input.
+   *       2) If set to false, the decoder simply produces a null in the output.
+   *    </li>
+   */
+  public HashTableDictionaryEncoder(D dictionary, boolean encodeNull) {
+    this(dictionary, encodeNull, SimpleHasher.INSTANCE);
+  }
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary.
+   * @param encodeNull a flag indicating if null should be encoded.
+    *     It determines the behaviors for processing null values in the input during encoding/decoding.
+    *    <li>
+    *      For encoding, when a null is encountered in the input,
+    *      1) If the flag is set to true, the encoder searches for the value in the dictionary,
+    *      and outputs the index in the dictionary.
+    *      2) If the flag is set to false, the encoder simply produces a null in the output.
+    *    </li>
+    *    <li>
+    *      For decoding, when a null is encountered in the input,
```

Comment:
not decoding happens here?

Reply:
The descriptions about decoding are removed. Thanks.